src/hpr_multi_buy.erl

```diff
@@ -32,38 +32,30 @@ init() ->
 update_counter(_Key, Max) when Max =< 0 ->
     {error, ?MAX_TOO_LOW};
 update_counter(Key, Max) ->
-    case application:get_env(hpr, multi_buy_enabled, true) of
-        false ->
-            case
-                ets:update_counter(
-                    ?ETS, Key, {2, 1}, {default, 0, erlang:system_time(millisecond)}
-                )
-            of
-                C when C > Max ->
-                    {error, ?MULTIBUY};
-                _C ->
-                    {ok, false}
-            end;
-        true ->
+    case
+        ets:update_counter(
+            ?ETS, Key, {2, 1}, {default, 0, erlang:system_time(millisecond)}
+        )
+    of
+        LocalCounter when LocalCounter > Max ->
+            {error, ?MULTIBUY};
+        LocalCounter ->
             case request(Key) of
-                {ok, C} when C > Max ->
+                {ok, ServiceCounter} when ServiceCounter > Max ->
+                    ets:update_counter(
+                        ?ETS,
+                        Key,
+                        {2, Max - LocalCounter + 1},
+                        {default, 0, erlang:system_time(millisecond)}
+                    ),
```

Comment:
```suggestion
                    ets:update_element(
                        ?ETS,
                        Key,
                        {2, ServiceCounter}
                    ),
```
`update_element/3` helps make it clearer that the next time around this packet will contain a value known to be higher than the max. And we don't have to do math 😄

Reply:
`update_element/3`  is not `atomic and isolated`